tls/s2n_resume.c

```diff
@@ -767,6 +767,32 @@ struct s2n_ticket_key *s2n_find_ticket_key(struct s2n_config *config, const uint
     return NULL;
 }
 
+struct s2n_unique_ticket_key {
+    uint8_t initial_key[S2N_AES256_KEY_LEN];
+    uint8_t info[S2N_AES256_KEY_LEN];
+    uint8_t output_key[S2N_AES256_KEY_LEN];
+};
+
+static S2N_RESULT s2n_resume_generate_unique_ticket_key(struct s2n_unique_ticket_key *key)
+{
+    RESULT_ENSURE_REF(key);
+
+    struct s2n_blob in_key_blob = { 0 };
+    RESULT_GUARD_POSIX(s2n_blob_init(&in_key_blob, key->initial_key, sizeof(key->initial_key)));
+    struct s2n_blob out_key_blob = { 0 };
+    RESULT_GUARD_POSIX(s2n_blob_init(&out_key_blob, key->output_key, sizeof(key->output_key)));
+    struct s2n_blob info_blob = { 0 };
+    RESULT_GUARD_POSIX(s2n_blob_init(&info_blob, key->info, sizeof(key->info)));
+    struct s2n_blob salt = { 0 };
+    RESULT_GUARD_POSIX(s2n_blob_init(&salt, NULL, 0));
+
+    DEFER_CLEANUP(struct s2n_hmac_state hmac = { 0 }, s2n_hmac_free);
+    RESULT_GUARD_POSIX(s2n_hmac_new(&hmac));
+    RESULT_GUARD_POSIX(s2n_hkdf(&hmac, S2N_HMAC_SHA256, &salt, &in_key_blob, &info_blob, &out_key_blob));
```

Comment:
Should we be using the common hmac workspace? I vaguely remember that allocating hmacs is expensive, and we have an issue to not do that in other places: https://github.com/aws/s2n-tls/issues/3206 Might at least be worth a TODO

Reply:
Can you tell me more about this? Is it the s2n_hmac_new call that is expensive? I remember @goatgoose did some work to remove hmac initialization when it wasn't needed, but IIRC that was done on every send/recv call so the codepath is way hotter.

Reply:
It was the s2n_hmac_new call. But it's been a while, so I'm not sure I remember the specifics. I think it was the openssl allocations?

Reply:
Yes - https://github.com/aws/s2n-tls/pull/4539 optimized s2n_send. This also likely made more of a difference since the hmac calls were unnecessarily invoked multiple times per send.

I'm not aware of anything in s2n_hmac_new being expensive other than the openssl allocations. Though, from the flamegraph I'm looking at, it seems like allocating the state would have similar performance to reusing the state due to a call to memmove in s2n_hmac_reset. However, allocating new state also requires freeing it, so reusing the state is probably cheaper overall. We'd probably need to generate more detailed performance data to know for sure though.

Reply:
I added a todo.